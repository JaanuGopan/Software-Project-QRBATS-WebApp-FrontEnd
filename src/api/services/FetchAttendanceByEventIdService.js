// FetchEventsService.js
import axios from "axios";
import ApiConstants from "../config/ApiConstants";

class FetchAttendanceByEventIdService {
  static async fetchAttendance(eventId) {
    try {
      const response = await axios.post(
        ApiConstants.getAttendanceByEventIdUrl,
        {
          eventId: eventId,
        }
      );

      return response.data;
    } catch (error) {
      console.error("Error fetching attendance:", error);
      throw error; // Re-throw the error to handle it in the component
    }
  }
}

export default FetchAttendanceByEventIdService;
