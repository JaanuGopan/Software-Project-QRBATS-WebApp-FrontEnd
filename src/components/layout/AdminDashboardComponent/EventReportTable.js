import React, { useState, useEffect } from "react";
import "../../../pages/AdminDashboard/AdminDashboard.css";
import { CiViewList } from "react-icons/ci";
import axios from "axios";
import FetchAttendanceByEventIdService from "../../../api/services/FetchAttendanceByEventIdService";

const EventReportTable = ({
  handleOpenReportWindow,
  search,
  onEventClick,
  eventList,
  attendedStudentList,
}) => {
  const [events, setEvents] = useState([]);
  const [selectedEvent, setSelectedEvent] = useState(null);
  const [attendanceList, setAttendanceList] = useState([]);

  useEffect(() => {
    // Update events whenever eventList prop changes
    setEvents(eventList);
  }, [eventList]);

  const handleEventClick = (event) => {
    setSelectedEvent(event);
    onEventClick(event);
  };

  const handleAttendanceList = (attendanceStudentList) => {
    setAttendanceList(attendanceStudentList);
    attendedStudentList(attendanceStudentList);
  };

  const handleReloadAttendanceList = async (eventId) => {
    FetchAttendanceByEventIdService.fetchAttendance(eventId)
      .then((attendanceList) => {
        setAttendanceList(attendanceList);
        handleAttendanceList(attendanceList);
        handleOpenReportWindow();
      })
      .catch((error) => {
        console.error("Error fetching attendance:", error);
      });
  };

  return (
    <div className="tableDesign">
      <table className="event-report-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Name</th>
            <th>Start Date</th>
            <th>End Date</th>
            <th>Venue</th>
            <th>Start Time</th>
            <th>End Time</th>
            <th>View</th>
          </tr>
        </thead>
        <tbody>
          {events
            .filter(
              (event) =>
                event.eventName.toLowerCase().includes(search.toLowerCase()) ||
                event.eventVenue.toLowerCase().includes(search.toLowerCase())
            )
            .map((event, index) => (
              <tr
                key={index}
                onClick={() => handleEventClick(event)}
                className={
                  selectedEvent === event ? "selected-row" : "event-row"
                }
              >
                <td>{index + 1}</td>
                <td>{event.eventName}</td>
                <td>{event.eventDate}</td>
                <td>{event.eventValidDate}</td>
                <td>{event.eventVenue}</td>
                <td>{event.eventTime}</td>
                <td>{event.eventEndTime}</td>
                <td>
                  <button
                    onClick={() => handleReloadAttendanceList(event.eventId)}
                    className="ViewButton"
                  >
                    <CiViewList className="EditIcon" />
                    <p className="ViewButtonLabel">View</p>
                  </button>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};

export default EventReportTable;
