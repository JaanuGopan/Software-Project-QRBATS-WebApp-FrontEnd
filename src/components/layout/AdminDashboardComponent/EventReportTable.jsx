import React, { useState, useEffect } from 'react';
import '../../../pages/AdminDashboard/AdminDashboard.css';
import { CiViewList } from 'react-icons/ci';
import axios from 'axios';
import FetchAttendanceByEventIdService from '../../../api/services/FetchAttendanceByEventIdService';
import { CircularProgress } from '@mui/material';

const EventReportTable = ({
  handleOpenReportWindow,
  search,
  onEventClick,
  eventList,
  attendedStudentList,
}) => {
  const [events, setEvents] = useState([]);
  const [selectedEvent, setSelectedEvent] = useState(null);
  const [attendanceList, setAttendanceList] = useState([]);

  useEffect(() => {
    // Update events whenever eventList prop changes
    setEvents(eventList);
  }, [eventList]);

  const handleEventClick = (event) => {
    setSelectedEvent(event);
    onEventClick(event);
  };

  const handleAttendanceList = (attendanceStudentList) => {
    setAttendanceList(attendanceStudentList);
    attendedStudentList(attendanceStudentList);
  };

  const handleReloadAttendanceList = async (eventId) => {
    await FetchAttendanceByEventIdService.fetchAttendance(eventId)
      .then((attendanceList) => {
        setAttendanceList(attendanceList);
        handleAttendanceList(attendanceList);
        handleOpenReportWindow();
      })
      .catch((error) => {
        console.error('Error fetching attendance:', error);
      });
  };

  const [loadingEventReport, setLoadingEventReport] = useState(false);
  const handleViewEventAttendance = async (eventId) => {
    try {
      setLoadingEventReport(true);
      await handleReloadAttendanceList(eventId);
    } finally {
      setLoadingEventReport(false);
    }
  };

  return (
    <div className="tableDesign">
      <table className="event-report-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Name</th>
            <th>Start Date</th>
            <th>Venue</th>
            <th>Start Time</th>
            <th>End Time</th>
            <th>View</th>
          </tr>
        </thead>
        {events.filter(
          (event) =>
            event.eventName.toLowerCase().includes(search.toLowerCase()) ||
            event.eventVenue.toLowerCase().includes(search.toLowerCase())
        ).length > 0 ? (
          <tbody>
            {events
              .filter(
                (event) =>
                  event.eventName.toLowerCase().includes(search.toLowerCase()) ||
                  event.eventVenue.toLowerCase().includes(search.toLowerCase())
              )
              .map((event, index) => (
                <tr
                  key={index}
                  onClick={() => handleEventClick(event)}
                  className={selectedEvent === event ? 'selected-row' : 'event-row'}
                >
                  <td>{index + 1}</td>
                  <td>{event.eventName}</td>
                  <td>{event.eventDate}</td>
                  <td>{event.eventVenue}</td>
                  <td>{event.eventTime}</td>
                  <td>{event.eventEndTime}</td>
                  <td>
                    {loadingEventReport ? (
                      <CircularProgress />
                    ) : (
                      <button
                        onClick={() => handleViewEventAttendance(event.eventId)}
                        className="ViewButton"
                      >
                        <CiViewList className="EditIcon" />
                        <p className="ViewButtonLabel">View</p>
                      </button>
                    )}
                  </td>
                </tr>
              ))}
          </tbody>
        ) : (
          <tbody className="nodata">No Data Available</tbody>
        )}
      </table>
    </div>
  );
};

export default EventReportTable;
