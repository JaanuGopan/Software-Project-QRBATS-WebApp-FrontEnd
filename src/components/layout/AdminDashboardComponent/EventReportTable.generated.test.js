import renderer from 'react-test-renderer';
import React, { useState, useEffect } from "react";

import { CiViewList } from "react-icons/ci";
import axios from "axios";
import FetchAttendanceByEventIdService from "../../../api/services/FetchAttendanceByEventIdService";
import EventReportTable from "./EventReportTable";

jest.mock("../../../pages/AdminDashboard/AdminDashboard.css");
jest.mock("react-icons/ci");
jest.mock("axios");
jest.mock("../../../api/services/FetchAttendanceByEventIdService");

const renderTree = tree => renderer.create(tree);
describe('<EventReportTable>', () => {
  it('should render component', () => {
    expect(renderTree(<EventReportTable 
    />).toJSON()).toMatchSnapshot();
  });
  
});