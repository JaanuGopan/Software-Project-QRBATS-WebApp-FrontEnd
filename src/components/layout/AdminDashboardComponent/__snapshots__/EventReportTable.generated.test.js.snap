// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventReportTable> should render component 1`] = `
<div
  className="tableDesign"
>
  <table
    className="event-report-tableArrangement"
  >
    <thead>
      <tr>
        <th>
          No
        </th>
        <th
          className="expand"
        >
          Name
        </th>
        <th>
          Start Date
        </th>
        <th>
          End Date
        </th>
        <th>
          Venue
        </th>
        <th>
          Start Time
        </th>
        <th>
          End Time
        </th>
        <th>
          View
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
